knowledge_repo/app/models.py

```diff
@@ -234,26 +235,59 @@ class Vote(db.Model):
 
 
 @unique_constructor(
-    lambda username: username,
-    lambda query, username: query.filter(User.username == username)
+    lambda identifier: identifier,
+    lambda query, identifier: query.filter(User.identifier == identifier)
 )
-class User(db.Model):
+class User(db.Model, UserMixin):
     __tablename__ = 'users'
 
     id = db.Column(db.Integer, primary_key=True)
-    username = db.Column(db.String(500))
     created_at = db.Column(db.DateTime, default=func.now())
 
+    identifier = db.Column(db.String(500))  # Unique identifier across all login methods
+
+    username = db.Column(db.String(500))  # Username used to log in (may differ from identifier)
+    password = db.Column(db.String(500))  # Password for local logins
+
+    name = db.Column(db.String(500))  # Name as determined by auth method
+    preferred_name = db.Column(db.String(500))  # Name as determined by user preferences
+
+    email = db.Column(db.String(500))  # Email address
+    avatar_uri = db.Column(db.Text())  # Either external url or data uri
+    active = db.Column(db.Boolean, default=True)
+
+    last_login_at = db.Column(db.DateTime)  # Date of last login
+
     _posts_assoc = db.relationship("PostAuthorAssoc")
     posts = association_proxy('_posts_assoc', 'post')  # This property should not directly modified
 
+    # Method overrides for the UserMixin class for flask_login
+    @property
+    def is_active(self):
+        return self.active
+
+    @property
+    def is_authenticated(self):
+        return True
+
+    @property
+    def is_anonymous(self):
+        return False
+
+    def get_id(self):
+        return self.identifier
+
+    can_logout = True
+
+    # Other useful methods
     @property
     def format_name(self):
-        username_to_name = current_repo.config.username_to_name
-        return username_to_name(self.username)
+        return self.preferred_name or self.name or self.identifier
+        # username_to_name = current_repo.config.username_to_name
```

Comment:
[question] how will this work with existing implementations that have a `username_to_name` function for formatting usernames? Are you planning on running a migration to backfill these columns?